### ISSUE TITLE
sort: slices.Sort is slightly slower than sort.Strings on some sorted inputs

### ISSUE THREAD

**Issue by qiulaidongfeng** [golang/go#60777]
### What version of Go are you using (`go version`)?

<pre>
$ go version devel go1.21-39effbc105 Fri Jun 9 04:07:29 2023 +0000 windows/amd64

</pre>

### Does this issue reproduce with the latest release?
yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
set GOOS=windows
set GOARCH=amd64
set GOAMD64=v2
</pre></details>

### What did you do?
Use the following benchmark code

<details><summary><code>code</code> </summary><br><pre>
package main

import (
	//"slices"
	"sort"
	"strconv"
	"testing"
)

var s []string

func init() {
	const n = 10000000
	s = make([]string, n)
	for i := 0; i < n; i++ {
		s[i] = strconv.Itoa(i)
	}
}

func BenchmarkStringSort(b *testing.B) {
	for i := 0; i < b.N; i++ {
		sort.Strings(s)
	}
}

// func BenchmarkStringSort(b *testing.B) {
// 	for i := 0; i < b.N; i++ {
// 		slices.Sort(s)
// 	}
// }

</pre></details>

### What did you expect to see?

Expected document to be successfully validated

Note: consider using the new slices.Sort function, which runs faster

### What did you see instead?
<code>
             │  sort.txt   │             slices.txt             │
             │   sec/op    │   sec/op     vs base               │
StringSort-4   123.7m ± 2%   134.0m ± 1%  +8.36% (p=0.002 n=10)
</code>

The test results do not match the document, which says slices is faster, but the test found that sort is faster.

**Comment by randall77** [comment]
> The test results do not match the document, which says slices is faster, but the test found that sort is faster.

Which document are you referring to?

**Comment by ianlancetaylor** [comment]
CC @eliben

**Comment by ianlancetaylor** [comment]
That's not a great benchmark, because it just calls the sorting function over and over again on the same slice.  After the first call, the slice will be fully sorted.  So you aren't really testing how long it takes to sort, you are testing how long it takes to determine that the slice is already sorted.

There are some existing benchmarks in slices/sort_benchmark_test.go and sort/sort_test.go that don't have that problem.

**Comment by qiulaidongfeng** [comment]
Document in https://pkg.go.dev/sort@master#Strings
I use the following benchmarks to sort different slices in each loop：
<details>
<summary>code</summary>
package main

import (
	//"slices"
	"sort"
	"strconv"
	"testing"
)

const n = 100000

func BenchmarkStringSort(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s := make([]string, n)
		for i := 0; i < n; i++ {
			s[i] = strconv.Itoa(i)
		}
		sort.Strings(s)
	}
}

// func BenchmarkStringSort(b *testing.B) {
// 	for i := 0; i < b.N; i++ {
// 		s := make([]string, n)
// 		for i := 0; i < n; i++ {
// 			s[i] = strconv.Itoa(i)
// 		}
// 		slices.Sort(s)
// 	}
// }

</details>

<code>
             │  sort.txt   │             slices.txt             │
             │   sec/op    │   sec/op     vs base               │
StringSort-4   38.63m ± 1%   36.23m ± 2%  -6.23% (p=0.000 n=10)
</code>

 Note: consider using the new slides. Sort function, which runs faster 

This document is easily understood as calling slices.Sort on the same slice (using [] string as an example) is faster than calling sort.Strings.

I think this part of the document should be changed to: "Note: Consider using the newer slices.Sort function to run faster on Not sorted slices", It's much less likely to be misunderstood.

**Comment by mitar** [comment]
> Which document are you referring to?

I think author means "documentation".

**Comment by eliben** [comment]
In my measurements, `slices.Sort` is 15-20% faster on a shuffled slice of strings (this is what I get by running the in-tree benchmarks @iant mentioned). For a presorted slice, I see measurement noise - either function is ~1% faster depending on the run.

Both functions end up in the same code-generated implementation, and the difference between them is the cost of dispatch (`StringSlice` interface for `sort.Strings` and generic dispatch for `cmp.Ordered` for `slices.Sort`) so I don't really see why sorting a sorted slice would be consistently faster.

@zhangyunhao116

**Comment by zhangyunhao116** [comment]
From this benchmark, the `sort.Strings` is slightly faster.
```
func sortedStrings() []string {
	const N = 10000
	x := make([]string, N)
	for i := 0; i < N; i++ {
		x[i] = strconv.Itoa(i)
	}
	slices.Sort(x)
	return x
}

func BenchmarkSlices(b *testing.B) {
	x := sortedStrings()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		slices.Sort(x)
	}
}

func BenchmarkSort(b *testing.B) {
	x := sortedStrings()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		sort.Strings(x)
	}
}
```
We can see the diff from its flame graph. The reason is that the slices use `cmp.Less`, which consumes too much CPU on the `isNaN`. IMOP, we can optimize this pattern in the compiler, there are many types like the `string` that don't require the `isNaN`.

<img width="1817" alt="Screenshot 2023-06-15 at 13 02 03" src="https://github.com/golang/go/assets/18065074/f69ccd85-55d8-41e8-9290-33031f6e5df9">

**Comment by bronze1man** [comment]
"IMOP, we can optimize this pattern in the compiler, there are many types like the string that don't require the isNaN."
Can golang just add compile time type check (rust called "[Zero Cost Abstractions](https://doc.rust-lang.org/beta/embedded-book/static-guarantees/zero-cost-abstractions.html)") to only check isNan? Or the compiler can transform runtime type check to compile time type check?

**Comment by randall77** [comment]
That just looks like a missing optimization. `x==x` should always return true for strings, and it doesn't.
```
func f(x string) bool {
	return x == x
}
```
That should compile to the equivalent of `return true`, but it doesn't. It ends up calling `memequal` (which does shortcut immediately to return true, but the call itself is still there).

**Comment by eliben** [comment]
> For this benchmark, the `sort.Strings` is slightly faster.
> 
> ```
> func sortedStrings() []string {
> 	const N = 10000
> 	x := make([]string, N)
> 	for i := 0; i < N; i++ {
> 		x[i] = strconv.Itoa(i)
> 	}
> 	slices.Sort(x)
> 	return x
> }
> 
> func BenchmarkSlices(b *testing.B) {
> 	x := sortedStrings()
> 	b.ResetTimer()
> 	for i := 0; i < b.N; i++ {
> 		slices.Sort(x)
> 	}
> }
> 
> func BenchmarkSort(b *testing.B) {
> 	x := sortedStrings()
> 	b.ResetTimer()
> 	for i := 0; i < b.N; i++ {
> 		sort.Strings(x)
> 	}
> }
> ```

Yes, this data is different so I see slightly different results. `sort.Strings` is faster by 1-2%

@randall77  makes a good point about how this can be fixed.

In the meantime, I'm not sure whether the doc comment should be modified. The new `slices.Sort` is certainly much faster on "real" inputs; it seems like for some subset of pre-sorted inputs it can be  slightly slower. We could qualify the comment from "runs faster" to "runs faster in most cases", but I'm not sure it's worth it.

**Comment by eliben** [comment]
I've updated the issue title to better reflect the current understanding. @randall77 's https://go.dev/cl/503795 fixes this, but we're not sure whether this fix is worth it for inclusion in 1.21 or will have to wait for 1.22

In the meantime, my inclination is not to do anything about the documentation, given the special circumstances where it's inaccurate (only for sorted inputs, and even then, only for *some* sorted inputs). Happy to hear other opinions
